azurerm/resource_arm_netapp_account.go

```diff
@@ -0,0 +1,280 @@
+package azurerm
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/netapp/mgmt/2019-06-01/netapp"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/response"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/validate"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/features"
+	aznetapp "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/netapp"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tags"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmNetAppAccount() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmNetAppAccountCreateUpdate,
+		Read:   resourceArmNetAppAccountRead,
+		Update: resourceArmNetAppAccountCreateUpdate,
+		Delete: resourceArmNetAppAccountDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: aznetapp.ValidateNetAppAccountName,
+			},
+
+			"resource_group_name": azure.SchemaResourceGroupNameForDataSource(),
+
+			"location": azure.SchemaLocation(),
+
+			"active_directory": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"dns_servers": {
+							Type:     schema.TypeList,
+							Required: true,
+							Elem: &schema.Schema{
+								Type:         schema.TypeString,
+								ValidateFunc: aznetapp.ValidateActiveDirectoryDNSName,
+							},
+						},
+						"domain": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: aznetapp.ValidateActiveDirectoryDomainName,
+						},
+						"smb_server_name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validate.NoEmptyStrings,
+						},
+						"username": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validate.NoEmptyStrings,
+						},
+						"password": {
+							Type:         schema.TypeString,
+							Required:     true,
+							Sensitive:    true,
+							ValidateFunc: validate.NoEmptyStrings,
+							DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+								maskedNew := strings.Replace(new, new, "****************", -1)
+								return (new == d.Get(k).(string)) && (maskedNew == old)
+							},
```

Comment:
@katbyte , I am not sure what you mean about "we should pull the value forward from config/state". Could you share some code sample for your suggestion for this? I know "make test" would fail if I removed this diff suppress function.

$ make testacc TEST=./azurerm TESTARGS="-parallel 11 -test.run=TestAccAzureRMNetApp"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./azurerm -v -parallel 11 -test.run=TestAccAzureRMNetApp -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMNetAppAccount
=== RUN   TestAccAzureRMNetAppAccount/update
=== RUN   TestAccAzureRMNetAppAccount/update/update
=== PAUSE TestAccAzureRMNetAppAccount/update/update
=== CONT  TestAccAzureRMNetAppAccount/update/update
=== RUN   TestAccAzureRMNetAppAccount/DataSource
=== RUN   TestAccAzureRMNetAppAccount/DataSource/basic
=== PAUSE TestAccAzureRMNetAppAccount/DataSource/basic
=== CONT  TestAccAzureRMNetAppAccount/DataSource/basic
=== RUN   TestAccAzureRMNetAppAccount/basic
=== RUN   TestAccAzureRMNetAppAccount/basic/basic
=== PAUSE TestAccAzureRMNetAppAccount/basic/basic
=== CONT  TestAccAzureRMNetAppAccount/basic/basic
=== RUN   TestAccAzureRMNetAppAccount/complete
=== RUN   TestAccAzureRMNetAppAccount/complete/complete
=== PAUSE TestAccAzureRMNetAppAccount/complete/complete
=== CONT  TestAccAzureRMNetAppAccount/complete/complete
--- FAIL: TestAccAzureRMNetAppAccount (436.81s)
    --- FAIL: TestAccAzureRMNetAppAccount/update (0.00s)
        --- FAIL: TestAccAzureRMNetAppAccount/update/update (115.72s)
            testing.go:569: Step 1 error: After applying this step, the plan was not empty:

                DIFF:

                UPDATE: azurerm_netapp_account.test
                  active_directory.#:                     "1" => "1"
                  active_directory.0.dns_servers.#:       "1" => "1"
                  active_directory.0.dns_servers.0:       "1.2.3.4" => "1.2.3.4"
                  active_directory.0.domain:              "westcentralus.com" => "westcentralus.com"
                  active_directory.0.id:                  "e6b944e9-fd12-4e92-12be-615da79a0137" => "e6b944e9-fd12-4e92-12be-615da79a0137"
                  active_directory.0.organizational_unit: "OU=FirstLevel" => "OU=FirstLevel"
                  active_directory.0.password:            "****************" => "aduserpwd"
                  active_directory.0.smb_server_name:     "SMBSERVER" => "SMBSERVER"
                  active_directory.0.username:            "aduser" => "aduser"
                  id:                                     "/subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070306264354/providers/Microsoft.NetApp/netAppAccounts/acctest-NetAppAccount-191019070306264354" => "/subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070306264354/providers/Microsoft.NetApp/netAppAccounts/acctest-NetAppAccount-191019070306264354"
                  location:                               "eastus2" => "eastus2"
                  name:                                   "acctest-NetAppAccount-191019070306264354" => "acctest-NetAppAccount-191019070306264354"
                  resource_group_name:                    "acctestRG-netapp-191019070306264354" => "acctestRG-netapp-191019070306264354"
                  tags.ENV:                               "Test" => "Test"



                STATE:

                azurerm_netapp_account.test:
                  ID = /subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070306264354/providers/Microsoft.NetApp/netAppAccounts/acctest-NetAppAccount-191019070306264354
                  provider = provider.azurerm
                  active_directory.# = 1
                  active_directory.0.dns_servers.# = 1
                  active_directory.0.dns_servers.0 = 1.2.3.4
                  active_directory.0.domain = westcentralus.com
                  active_directory.0.id = e6b944e9-fd12-4e92-12be-615da79a0137
                  active_directory.0.organizational_unit = OU=FirstLevel
                  active_directory.0.password = ****************
                  active_directory.0.smb_server_name = SMBSERVER
                  active_directory.0.username = aduser
                  location = eastus2
                  name = acctest-NetAppAccount-191019070306264354
                  resource_group_name = acctestRG-netapp-191019070306264354
                  tags.ENV = Test

                  Dependencies:
                    azurerm_resource_group.test
                azurerm_resource_group.test:
                  ID = /subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070306264354
                  provider = provider.azurerm
                  location = eastus2
                  name = acctestRG-netapp-191019070306264354
    --- PASS: TestAccAzureRMNetAppAccount/DataSource (0.00s)
        --- PASS: TestAccAzureRMNetAppAccount/DataSource/basic (103.02s)
    --- PASS: TestAccAzureRMNetAppAccount/basic (0.00s)
        --- PASS: TestAccAzureRMNetAppAccount/basic/basic (111.85s)
    --- FAIL: TestAccAzureRMNetAppAccount/complete (0.00s)
        --- FAIL: TestAccAzureRMNetAppAccount/complete/complete (106.22s)
            testing.go:569: Step 0 error: After applying this step, the plan was not empty:

                DIFF:

                UPDATE: azurerm_netapp_account.test
                  active_directory.#:                     "1" => "1"
                  active_directory.0.dns_servers.#:       "1" => "1"
                  active_directory.0.dns_servers.0:       "1.2.3.4" => "1.2.3.4"
                  active_directory.0.domain:              "westcentralus.com" => "westcentralus.com"
                  active_directory.0.id:                  "c116d2fa-297e-a04f-5e7d-97671e2f2ac9" => "c116d2fa-297e-a04f-5e7d-97671e2f2ac9"
                  active_directory.0.organizational_unit: "OU=FirstLevel" => "OU=FirstLevel"
                  active_directory.0.password:            "****************" => "aduserpwd"
                  active_directory.0.smb_server_name:     "SMBSERVER" => "SMBSERVER"
                  active_directory.0.username:            "aduser" => "aduser"
                  id:                                     "/subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070836852308/providers/Microsoft.NetApp/netAppAccounts/acctest-NetAppAccount-191019070836852308" => "/subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070836852308/providers/Microsoft.NetApp/netAppAccounts/acctest-NetAppAccount-191019070836852308"
                  location:                               "eastus2" => "eastus2"
                  name:                                   "acctest-NetAppAccount-191019070836852308" => "acctest-NetAppAccount-191019070836852308"
                  resource_group_name:                    "acctestRG-netapp-191019070836852308" => "acctestRG-netapp-191019070836852308"
                  tags.ENV:                               "Test" => "Test"



                STATE:

                azurerm_netapp_account.test:
                  ID = /subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070836852308/providers/Microsoft.NetApp/netAppAccounts/acctest-NetAppAccount-191019070836852308
                  provider = provider.azurerm
                  active_directory.# = 1
                  active_directory.0.dns_servers.# = 1
                  active_directory.0.dns_servers.0 = 1.2.3.4
                  active_directory.0.domain = westcentralus.com
                  active_directory.0.id = c116d2fa-297e-a04f-5e7d-97671e2f2ac9
                  active_directory.0.organizational_unit = OU=FirstLevel
                  active_directory.0.password = ****************
                  active_directory.0.smb_server_name = SMBSERVER
                  active_directory.0.username = aduser
                  location = eastus2
                  name = acctest-NetAppAccount-191019070836852308
                  resource_group_name = acctestRG-netapp-191019070836852308
                  tags.ENV = Test

                  Dependencies:
                    azurerm_resource_group.test
                azurerm_resource_group.test:
                  ID = /subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070836852308
                  provider = provider.azurerm
                  location = eastus2
                  name = acctestRG-netapp-191019070836852308
FAIL
FAIL    github.com/terraform-providers/terraform-provider-azurerm/azurerm       448.933s
make: *** [testacc] Error 1

Reply:
Please don't resolve resolve conversations until a consensus has been reached.

What i mean here is that instead of doing a diff suppress we should on read grab the old value from state/config and reapply it instead of reading it from the API (because it will always be ***)